### PR TITLE
CORTX-28044: Add and Remove Access Capability

### PR DESCRIPTION
**Pull request by abhijit1patil**
Signed-off-by: abhijit1patil <abhijit.1.patil@seagate.com>

# Problem Statement
- [CORTX-28044: Add a User Capability ](https://jts.seagate.com/browse/CORTX-28044)
- [CORTX-28045: Remove a User Capability ](https://jts.seagate.com/browse/CORTX-28045)
# Design
-  [Add a User Capability REST API Specification](https://seagate-systems.atlassian.net/l/c/N0iX7iVP)
- [Remove a User Capability REST API Specification](https://seagate-systems.atlassian.net/l/c/es4VV1D8)

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent
# Build and Deployments 

-  [Jenkins:- Custom Build Job ](https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5253/)
- Custom Build Image:-
1.  CORTX_ALL_IMAGE:- cortx-docker.colo.seagate.com/seagate/cortx-all:2.0.0-5253-custom-ci
2.  CORTX_SERVER_IMAGE:- cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-5253-custom-ci

- [Jenkins:- Deployment with Custom Build Job](https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/677/)
- Successfully Deployed VM
image

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM
# Dev Testing
Testcase 0:- Add Access Caps
image

Testcase 1:- Remove Access Caps
image

Testcase2:- Invalid access caps
image

Testcase3:- Invalid permission for caps , 
image

Testcase4:- False Positive scanario, if given input pattern is not followed then response will not get updated correctly
image

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/core/controllers/rgw/s3/users.py [hunk 1]
```diff
@@ -89,6 +89,13 @@ class RemoveKeySchema(S3BaseSchema):
     key_type = fields.Str(data_key=const.KEY_TYPE, missing=None,
                     validate=validate.OneOf(const.SUPPORTED_KEY_TYPES))
 
+class UserCapsSchema(S3BaseSchema):
+    """
+    S3 Create/Add Access Key schema validation class.
+    """
```
Comment (rohitkolapkar) [inline csm/core/controllers/rgw/s3/users.py:95]: Please update doc string.

File: csm/core/services/rgw/s3/users.py [hunk 2]
```diff
@@ -102,3 +102,23 @@ async def remove_key(self, **remove_key_body):
...
+        uid = request_body.get(const.UID)
+        Log.debug(f"Add User caps for S3 IAM user by uid = {uid}")
+        return await self.execute_request(const.ADD_USER_CAPS_OPERATION, **request_body)
+
+    @Log.trace_method(Log.DEBUG)
+    async def remove_user_caps(self, **request_body):
+        """
+        This method will remove user caps for a new S3 IAM user.
+        :param **request_body: User body kwargs
+        """
```
Comment (rohitkolapkar) [inline csm/core/services/rgw/s3/users.py:121]: doc string : new IAM user ?

File: csm/core/services/rgw/s3/users.py [hunk 3]
```diff
@@ -102,3 +102,23 @@ async def remove_key(self, **remove_key_body):
         uid = remove_key_body.get(const.UID)
         Log.debug(f"Removing key for S3 IAM user by uid = {uid}")
         return await self.execute_request(const.REMOVE_KEY_OPERATION, **remove_key_body)
+
+    @Log.trace_method(Log.DEBUG)
+    async def add_user_caps(self, **request_body):
+        """
+        This method will add user caps for a new S3 IAM user.
+        :param **request_body: User body kwargs
```
Comment (rohitkolapkar) [inline csm/core/services/rgw/s3/users.py:110]: doc string : new IAM user ?

File: csm/core/data/models/rgw.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ class RgwErrors(Enum):
     InternalError = "Internal server error."
     NoSuchUser = "User does not exist."
     NoSuchBucket = "Bucket does not exist."
-    NoSuchKey = "No such access key."
+    NoSuchKey = "No such key."
```
Comment (rohitkolapkar) [inline csm/core/data/models/rgw.py:34]: in which scenario, s3 server returns - NoSuchKey ?

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/core/controllers/rgw/s3/users.py (hunk 1 above)
Comment (abhijit1patil) [reply csm/core/controllers/rgw/s3/users.py:95]: updated

File: csm/core/data/models/rgw.py (hunk 4 above)
Comment (abhijit1patil) [reply csm/core/data/models/rgw.py:34]: actually NoSuchKey is returned from bucket link operation.. if we give incorrect bucket name.. this issue pointed out by Amit

Comment (rohitkolapkar) [reply csm/core/data/models/rgw.py:34]: got it.

File: csm/core/services/rgw/s3/users.py [hunk 5]
```diff
@@ -102,3 +102,23 @@ async def remove_key(self, **remove_key_body):
...
         return await self.execute_request(const.REMOVE_KEY_OPERATION, **remove_key_body)
+
+    @Log.trace_method(Log.DEBUG)
+    async def add_user_caps(self, **request_body):
+        """
+        This method will add user caps to S3 IAM user.
+        :param **request_body: User body kwargs
+        """
+        uid = request_body.get(const.UID)
+        Log.debug(f"Add User caps S3 IAM user = {uid}")
```
Comment (pawankumarsrivastava) [inline csm/core/services/rgw/s3/users.py:113]: Make this Log.Info and change "Add User caps S3 IAM user" -> "Adding user capabilities for {uid}"

File: csm/core/services/rgw/s3/users.py [hunk 6]
```diff
@@ -102,3 +102,23 @@ async def remove_key(self, **remove_key_body):
         uid = remove_key_body.get(const.UID)
         Log.debug(f"Removing key for S3 IAM user by uid = {uid}")
         return await self.execute_request(const.REMOVE_KEY_OPERATION, **remove_key_body)
+
+    @Log.trace_method(Log.DEBUG)
+    async def add_user_caps(self, **request_body):
+        """
+        This method will add user caps to S3 IAM user.
```
Comment (pawankumarsrivastava) [inline csm/core/services/rgw/s3/users.py:109]: Remove S3. Write only IAM Users

File: csm/core/services/rgw/s3/users.py [hunk 7]
```diff
@@ -102,3 +102,23 @@ async def remove_key(self, **remove_key_body):
...
+        return await self.execute_request(const.ADD_USER_CAPS_OPERATION, **request_body)
+
+    @Log.trace_method(Log.DEBUG)
+    async def remove_user_caps(self, **request_body):
+        """
+        This method will remove user caps for a new S3 IAM user.
+        :param **request_body: User body kwargs
+        """
+        uid = request_body.get(const.UID)
+        Log.debug(f"remove User caps S3 IAM user = {uid}")
```
Comment (pawankumarsrivastava) [inline csm/core/services/rgw/s3/users.py:123]: Same as above

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/services/rgw/s3/users.py (hunk 5 above)
Comment (abhijit1patil) [reply csm/core/services/rgw/s3/users.py:113]: updated

File: csm/core/services/rgw/s3/users.py (hunk 6 above)
Comment (abhijit1patil) [reply csm/core/services/rgw/s3/users.py:109]: updated

File: csm/core/services/rgw/s3/users.py (hunk 7 above)
Comment (abhijit1patil) [reply csm/core/services/rgw/s3/users.py:123]: updated

File: csm/core/services/rgw/s3/users.py (hunk 2 above)
Comment (abhijit1patil) [reply csm/core/services/rgw/s3/users.py:121]: updated

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by rohitkolapkar — APPROVED** [review]

**Pull request merged**